[Serialization] Give swiftdocs a stable version

jrose-apple opened the pull request:
We're committing to this as a forwards-compatible format, and in most cases probably backwards-compatible as well! Break it apart from the AST/SIL swiftmodule format, give it its own version number, and…probably not have any problems, because there's nothing complicated here.

jrose-apple committed:
[Serialization] Pull swiftdoc serialization out to a separate file

No functionality change.

jrose-apple committed:
[Serialization] Split doc serialization out of the Serializer class

Extract the common utilities into a SerializerBase class, and make a
new DocSerializer.

jrose-apple committed:
[Serialization] Re-order block IDs to keep the doc blocks at the end

jrose-apple committed:
[Serialization] Preparation for giving swiftdoc its own version

The functionality change in this commit is that the control block in a
swiftdoc file is validated rather than just being ignored. Tests in
following commit.

jrose-apple committed:
[Serialization] Give swiftdocs a stable version

We're committing to this as a forwards-compatible format, and in most
cases probably backwards-compatible as well!

nkcsgexi commented:
Could you provide more context why do we need this version number while it's both forward and backward compatible?

jrose-apple commented:
`SWIFTDOC_VERSION_MINOR` isn't really important for forwards/backwards-compatibility, but it gives you the option to conditionalize behavior based on which version of the compiler generated the file. (For example, maybe in a newer format, an absent record indicates that a decl should be omitted from the generated interface, while in the older format it just means it has no comment.)

`SWIFTDOC_VERSION_MAJOR` is important in case you really do have to break compatibility some day. I suspect it won't actually go up at all, but if it does then how the rest of the file is processed can be conditionalized on the major version number.

This format is basically [semantic versioning](https://semver.org), but a file format isn't exactly the same thing as an API, so I didn't mention semver by name in the doc comment.

nkcsgexi commented:
Thanks for the clarification! Can we derive `SWIFTDOC_VERSION_MINOR` from the complier/module format version we're using rather than maintaining it separately? Does this change also imply that all tooling invocation should also specify `SWIFTDOC_VERSION_MAJOR`?

akyrtzi commented on lib/Serialization/DocFormat.h:
```diff
@@ -0,0 +1,85 @@
+///
+/// Be very careful when changing this block; it must remain stable. Adding new
+/// records is okay---they will be ignored---but modifying existing ones must be
+/// done carefully. You may need to update the version when you do so.
```
What is the effect of adding a new 'field' at the end of an existing record, could you clarify it in the comments ?
Can we determine how many fields a record has at deserialization so we don't have to update and check for a version number ?

jrose-apple replied:
You can kind of do that—see https://github.com/apple/swift/blob/bae3964c07a976ee0978c0c2400e23e3252e60e8/lib/Serialization/ModuleFile.cpp#L200-L216—but BCRecordLayout doesn't support it. You have to read records manually in that case. Adding new records isn't really that expensive, though, so I don't think it's too much of a problem to do that.

For a stable format, the version number is supposed to be higher-level than just "did I change something". If it makes no difference, it's fine not to update it (and if you really don't want to use it, we could just drop the "minor" version entirely).

akyrtzi replied:
I think it's good that you added the minor version, as an extra thing we may find useful to check,  I'm mainly trying to understand if we should update it when adding a new field.

Adding a new field would force us to switch to fully manually reading the record ? Would we be able to do a check and switch to using a different BCRecordLayout to deserialize that record, and could this check occur without needing to update the minor version ?

Pointing in the documentation to the example that you gave on how to manually check for the number of fields would be useful.

jrose-apple replied:
It's unfortunately worse than that: adding a new field will **break existing tools**, because BCRecordLayout currently asserts that the number of fields matches exactly. https://github.com/apple/swift-llvm/blob/stable/include/llvm/Bitcode/RecordLayout.h#L261

We could change that, of course, or add a new API `readIgnoringTrailingFields` or something, but the point is that you can only add new fields if there aren't existing tools expecting a fixed number of fields.

akyrtzi commented on lib/Serialization/DocFormat.h:
```diff
@@ -0,0 +1,85 @@
+///
+/// Increment this value when making a backwards-compatible change that might
+/// be interesting to test for. However, if old swiftdoc files are fully
+/// compatible with the new change, you do not need to increment this.
```
I'm probably reading this paragraph wrong but it seems to contradict itself, could you rephrase for clarify, are you making a distinction between "backwards-compatible" and "fully compatible" ? If yes I'd recommend to define in details (and maybe with some examples) what you mean by each.

jrose-apple commented:
> Thanks for the clarification! Can we derive `SWIFTDOC_VERSION_MINOR` from the complier/module format version we're using rather than maintaining it separately?

Unfortunately no. The goal here is to have a version number for a format that's stable across compiler versions. The binary swiftmodule format definitely isn't.

> Does this change also imply that all tooling invocation should also specify `SWIFTDOC_VERSION_MAJOR`?

I'm not quite sure what you're asking here. The way the version checks are written, all our tools are compiled with exactly one expected major version for any swiftdoc they read. If that version ever changes, we'll have to decide how the new tools handle the differing major versions, but the old tools will correctly reject the newer generated files.

(It's also theoretically possible to decide what version to use when *generating* the file—say, if a user doesn't use any complicated features—but that's probably not something you'll need, especially since the way your deserialization is written will already ignore records it doesn't understand.)

akyrtzi commented:
@jrose-apple thanks so much for doing this! My main concern is for `DocFormat.h` to contain documentation with crystal clear instructions and directions on how to make changes and maintain compatibility, it should have no ambiguities.

nkcsgexi commented:
> I'm not quite sure what you're asking here. The way the version checks are written, all our tools are compiled with exactly one expected major version for any swiftdoc they read. If that version ever changes, we'll have to decide how the new tools handle the differing major versions, but the old tools will correctly reject the newer generated files.

ah, so the versioning is entirely transparent to compiler clients like sourcekitd, and its purpose is to separate the evolution of doc-serialization with compiler and module format because in practice it's way more stable, right?

jrose-apple commented:
> ah, so the versioning is entirely transparent to compiler clients like sourcekitd, and its purpose is to separate the evolution of doc-serialization with compiler and module format because in practice it's way more stable, right?

Yep, that sounds right to me!

> My main concern is for `DocFormat.h` to contain documentation with crystal clear instructions and directions on how to make changes and maintain compatibility, it should have no ambiguities.

Agreed. I'll add more comments based on the discussions we're having now.

jrose-apple committed:
[Serialization] Avoid using BCRecordLayout for decoding swiftdocs

BCRecordLayout currently assumes that the layout described in source
always matches the layout in the bitstream being read. Since we want
swiftdocs to be a forward-compatible format, avoid using it for
deserialization.

(In the future, we may want to augment BCRecordLayout to handle
records with more fields than expected. For now, though, this is a
sufficient change.)

jrose-apple committed:
Docs on how to make backwards-compatible changes using LLVM bitstream

jrose-apple commented:
@akyrtzi @nkcsgexi How does this look?

@swift-ci Please smoke test

akyrtzi reviewed (approved):
Love the detailed documentation!

jrose-apple commented:
@swift-ci Please test
